privacysandbox/privacysandbox.go

```diff
@@ -0,0 +1,186 @@
+package privacysandbox
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+)
+
+type Topic struct {
+	SegTax   int
+	SegClass string
+	SegIDs   []int
+}
+
+// ParseTopicsFromHeader parses the Sec-Browsing-Topics header data into Topics object
+func ParseTopicsFromHeader(secBrowsingTopics string) []Topic {
+	var topics []Topic
+
+	for _, seg := range strings.Split(secBrowsingTopics, ",") {
+		if topic, ok := parseTopicSegment(seg); ok {
+			topics = append(topics, topic)
+		}
+
+		if len(topics) == 10 {
+			break
+		}
+	}
+
+	return topics
+}
+
+// parseTopicSegment parses a single topic segment from the header into Topics object
+func parseTopicSegment(seg string) (Topic, bool) {
+	seg = strings.TrimSpace(seg)
+	if seg == "" || strings.HasPrefix(seg, "();p=") {
+		return Topic{}, false
+	}
+
+	segment := strings.Split(seg, ";")
+	if len(segment) != 2 {
+		return Topic{}, false
+	}
+
+	segmentsIDs := strings.TrimSpace(segment[0])
+	if len(segmentsIDs) < 3 || segmentsIDs[0] != '(' || segmentsIDs[len(segmentsIDs)-1] != ')' {
+		return Topic{}, false
+	}
+
+	segtax, segclass := parseSegTaxSegClass(segment[1])
+	if segtax == 0 || segclass == "" {
+		return Topic{}, false
+	}
+
+	return Topic{
+		SegTax:   segtax,
+		SegClass: segclass,
+		SegIDs:   parseSegmentIDs(segmentsIDs[1 : len(segmentsIDs)-1]),
+	}, true
+}
+
+func parseSegTaxSegClass(seg string) (int, string) {
+	taxanomyModel := strings.Split(seg, ":")
+	if len(taxanomyModel) != 3 {
+		return 0, ""
+	}
+
+	// taxanomyModel[0] is v=browser_version, we don't need it
+	taxanomyVer := strings.TrimSpace(taxanomyModel[1])
+	taxanomy, err := strconv.Atoi(taxanomyVer)
+	if err != nil || taxanomy < 1 || taxanomy > 10 {
+		return 0, ""
+	}
+
+	segtax := 600 + (taxanomy - 1)
+	segclass := strings.TrimSpace(taxanomyModel[2])
+	return segtax, segclass
+}
+
+// parseSegmentIDs parses the segment ids from the header string into int array
+func parseSegmentIDs(segmentsIDs string) []int {
+	var selectedSegmentIDs []int
+	for _, segmentID := range strings.Fields(segmentsIDs) {
+		segmentID = strings.TrimSpace(segmentID)
+		if selectedSegmentID, err := strconv.Atoi(segmentID); err == nil && selectedSegmentID > 0 {
+			selectedSegmentIDs = append(selectedSegmentIDs, selectedSegmentID)
+		}
+	}
+
+	return selectedSegmentIDs
+}
+
+func UpdateUserDataWithTopics(userData []openrtb2.Data, headerData []Topic, topicsDomain string) []openrtb2.Data {
+	if topicsDomain == "" {
+		return userData
+	}
+
+	headerDataMap := createHeaderDataMap(headerData)
+
+	for i, data := range userData {
+		ext := &Topic{}
+		err := json.Unmarshal(data.Ext, ext)
+		if err != nil {
+			continue
+		}
+
+		if ext.SegTax == 0 || ext.SegClass == "" {
+			continue
+		}
+
+		if newSegIDs := mergeSegIDs(data.Name, topicsDomain, *ext, data.Segment, headerDataMap); newSegIDs != nil {
```

Comment:
Added comment and unit test `user data with duplicate topic`

```
	// headerDataMap groups segIDs by segtax and segclass for faster lookup and tracking of new segIDs yet to be added to user.data
	// tracking is done by removing segIDs from segIDsMap once they are added to user.data, ensuring that headerDataMap will always have unique segtax-segclass-segIDs
	// the only drawback of tracking via deleting segtax-segclass from headerDataMap is that this would not track duplicate entries within user.data which is fine because we are only merging header data with the provided user.data
	headerDataMap := createHeaderDataMap(headerData)
```